### DISCUSSION TITLE
Observability & Instrumentation

### DISCUSSION THREAD

**Discussion by jkratz55** [earthboundkid/requests#56] [Support] [answered]
We are currently using a very thin wrapper around the standard http.Client and capture metrics using Prometheus for response times, errors, etc., when making outbound HTTP requests. The way we implemented it was by adding Before and After hooks into our wrapper type. 

Looking at this library, it looks like we'd need to implement the RoundTripper interface wrapper an http.Transport and do our instrumentation there and set the Transport on the builder. Is that a reasonable approach using this library? Curios if anyone else has taken a similar approach.

**Comment by earthboundkid** [comment]
Using a `requests.RoundTripFunc` that captures Prometheus metrics as a Transport sounds like the right approach. I do some very minor logging using a `RoundTripFunc` here: https://github.com/spotlightpa/almanack/blob/master/pkg/common/httplogger.go#L11-L20

If you have a snippet using Prometheus, it might be a good [addition to the wiki](https://github.com/carlmjohnson/requests/wiki).

**Reply by jkratz55** [reply]
Perfect! I'm currently evaluating the use of this library within our ecosystem and will share a snippet when I get something together.

**Comment by earthboundkid** [comment, marked as answer]
I want to add this to requests soon. Something simple, just `func(base http.RoundTripper, fn func(req *http.Request, res *http.Response, err error, duration time.Duration)) Transport` that you can fill in the details for.

**Reply by earthboundkid** [reply]
See #65.
